### ISSUE TITLE
Webpack generating nondeterministic hashes, seems to not merge duplicate chunks

### ISSUE THREAD

**Issue by nanramakr** [webpack/webpack#17009]
We are facing an issue where the contenthash for entry points in our project changes between builds, without any source file changes.
.
Eg. with this config-
```
output: {
    publicPath: '/static/bundles/',
    filename: '[contenthash:8]-[name].js',
    chunkFilename: '[contenthash:8]-[name].js',
},
```
For entrypoint `abc`, build 1 generated `1b2c3d4e-abc.js`, build 2 may generate `f5a6b7c8-abc.js`

On introspecting these bundles, and setting `optimization.chunkIds` to `named`, it seems to occur because of a single chunk, duplicated several times. The contents of the file are the same, only the succeeding hash is different (and interestingly seems to be incrementing). As a result the bundle contenthash is changing.

<img width="1598" alt="Screenshot 2023-04-16 at 11 52 46 PM" src="https://user-images.githubusercontent.com/8663117/232407138-327ae46a-c3ea-4166-a2e6-9698eab83ad0.png">

On checking the contents of this vendor chunk, it looks like there are several such chunks that have been created.
```
(self.webpackChunk = self.webpackChunk || []).push([
  [
    "vendors-node_modules_lodash__baseIsEqual_js-node_modules_lodash_get_js-node_modules_react-lif-22f2913",
    "vendors-node_modules_lodash__baseIsEqual_js-node_modules_lodash_get_js-node_modules_react-lif-b9b55d0",
    "vendors-node_modules_lodash__baseIsEqual_js-node_modules_lodash_get_js-node_modules_react-lif-22f2910",
    "vendors-node_modules_lodash__baseIsEqual_js-node_modules_lodash_get_js-node_modules_react-lif-22f2911",
    "vendors-node_modules_lodash__baseIsEqual_js-node_modules_lodash_get_js-node_modules_react-lif-22f2912",
    "vendors-node_modules_lodash__baseIsEqual_js-node_modules_lodash_get_js-node_modules_react-lif-22f2914",
    "vendors-node_modules_lodash__baseIsEqual_js-node_modules_lodash_get_js-node_modules_redux-dev-d37e7d",
    "vendors-node_modules_lodash__baseIsEqual_js-node_modules_lodash_get_js-node_modules_react-lif-b9b55d1",
    "vendors-node_modules_lodash__baseIsEqual_js-node_modules_lodash_get_js-node_modules_react-lif-b9b55d2",
    "vendors-node_modules_lodash__baseIsEqual_js-node_modules_lodash_get_js-node_modules_react-lif-b9b55d3",
  ],
```

Webpack optimization config is as follows, and we are running production mode:
```
optimization: {
        nodeEnv: 'production',
        minimize: true,
        removeAvailableModules: false,
        removeEmptyChunks: true,
        chunkIds: 'named',
        minimizer: [
            new TerserPlugin({
                parallel: true,
                terserOptions: {
                    ecma: 5,
                    ie8: false,
                    safari10: false,
                },
            }),
        ],
    },
```

Continuing to investigate on our end, but any help would be appreciated! Ideally only one such chunk should exist for this group of modules, but there seems to be 5. One of the big questions that I had about these named chunks was what is the 7 character hash at the end of the chunk name based on, and what could cause it to increment?

Webpack version: 5.79.0
Node version: 16.13.2
OS version: mac OS Ventura (Apple M1)

**Comment by alexander-akait** [comment]
> Continuing to investigate on our end, but any help would be appreciated! Ideally only one such chunk should exist for this group of modules, but there seems to be 5. One of the big questions that I had about these named chunks was what is the 7 character hash at the end of the chunk name based on, and what could cause it to increment?

Do you check it on the same OS?
Can you show me full configuration? If no can you show me `splitChunks` configuration
Also do you have any runtime values in `DefinePlugin`?

**Comment by alexander-akait** [comment]
Please remove `loader: 'cache-loader',` it potential breaks your cache and many other things, we have built-in cache https://webpack.js.org/configuration/cache/#cachetype

**Comment by alexander-akait** [comment]
Also please avoid using `worker-loader` in favor `new Worker(...)` syntax , `file-loader` is deprecated in favor https://webpack.js.org/guides/asset-modules/, your have a lot of deprecated and old things, and all of them can be a problem...

**Comment by alexander-akait** [comment]
For debug purposes please try to disable `thread-loader`

**Comment by nanramakr** [comment]
> For debug purposes please try to disable `thread-loader`

Hi @alexander-akait thank you for the suggestions! I have removed the use of thread-loader, worker-loader and cache-loader. Instead of cache-loader, now we are using the webpack inbuilt cache

```
module.exports = {
    cache: true,
}
```
And still running into the same issues. Removing file-loader caused a huge diff between subsequent builds, so left that in for now.

I had a question, given a named chunk such as `vendors-node_modules_lodash__baseIsEqual_js-node_modules_lodash_get_js-node_modules_react-lif-b9b55d1`, how does the succeeding hash `b9b55d1` get derived? I wanted to understand why it is incrementing to help debug this issue. I first thought this may be the module id, but even setting `"moduleIds": "deterministic"`  didn't seem to work

**Comment by alexander-akait** [comment]
We have a plugin - ReadContentHash to generate real content hash, you can try to disable it https://webpack.js.org/configuration/optimization/#optimizationrealcontenthash, but I think it will not help, not sure, but looks like some of loader/plugins change code/requests, so it happens, hard to say, need debug with real code

**Comment by alexander-akait** [comment]
Try to disable `worker-loader` for debug purposes

**Comment by ArcanoxDragon** [comment]
We are seeing this at my company in our primary Webpack build process as well. Interestingly, I've been trying to find a consistent way to reproduce it, and I'm not having any luck. I managed to get module IDs to be different between three subsequent builds on the same source, but every subsequent build after that has yielded the exact same results (even if I wipe out the cache and the output directory). It's almost as if it's somehow time-based, or there's some other "random" influence.

**Comment by ArcanoxDragon** [comment]
I have discovered some quirky behavior while trying to work around this issue in my company's Webpack project. It seems the _ordering_ of the keys in the `entry` object given in the Webpack config can change the output, specifically the module IDs. We derive the entry points dynamically at compile time. Our project structure is something like the following:

```
webpack.config.ts
projects/
    a/
        scripts/
            entry-point-a.ts
        src/
            moduleOne.ts
        styles/
            _partial-a.scss
            stylesheet-a.scss
    b/
        scripts/
            entry-point-b.ts
        src/
            moduleTwo.ts
        styles/
            _partial-b.scss
            stylesheet-b.scss
```

The files in each project's `scripts` and `styles` folders are discovered as entry points using [`globby`](/sindresorhus/globby). We turn the array of files returned by `globby` into a map of entry points to pass to the `entry` option on our Webpack config. The entry object looks something like this:

```js
{
    "a/entry-point-a": "./projects/a/scripts/entry-point-a.ts",
    "b/entry-point-b": "./projects/b/scripts/entry-point-b.ts",
    "styles/a/stylesheet-a": "./projects/a/styles/stylesheet-a.scss",
    "styles/b/stylesheet-b": "./projects/b/styles/stylesheet-b.scss",
}
```

I have been struggling to reproduce this "non-deterministic build" problem locally, but it occurs quite consistently on our automated build server. Purely on a random hunch, I decided to try _shuffling_ the ordering of the keys in the `entry` object I give to Webpack (to simulate if `globby` gave us files back in a different order than last time). Sure enough, the output changed completely. Certain things that were mangled became non-mangled and vice versa, and the numeric module IDs got shuffled around. Every single build I ran with shuffled input was different.

I then took out the shuffling code and instead did a `.sort()` on the array I get back from `globby`. The build output is back to being deterministic locally, though I have not tried this fix on our build server yet.

@nanramakr: does your Webpack config also use dynamically-derived entry points? What happens if you try sorting the entry object keys in a deterministic order?
